[16.01] Always install a specific version of Galaxy pip in common_startup.sh.

natefoo opened the pull request:
Pinning the version should avoid problems related to future pip releases. It does mean that any fixes we need to release for Galaxy pip won't automatically be received, but that's probably the way it should be. This also upgrades from 8.0.0 to 8.0.2. 8.0.2+gx1 wheels are up on wheels.galaxyproject.org.

Fixes #1544 again.

related: galaxyproject/ansible-galaxy#9

natefoo committed:
Always install a specific version of Galaxy pip in common_startup.sh.

martenson commented:
The failing tests are relevant I think. 

```
Traceback (most recent call last):
  File "/galaxy/.venv/local/lib/python2.7/site-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/galaxy/.venv/local/lib/python2.7/site-packages/pip/commands/install.py", line 257, in run
    InstallRequirement.from_line(name, None))
  File "/galaxy/.venv/local/lib/python2.7/site-packages/pip/req.py", line 172, in from_line
    return cls(req, comes_from, url=url, prereleases=prereleases)
  File "/galaxy/.venv/local/lib/python2.7/site-packages/pip/req.py", line 70, in __init__
    req = pkg_resources.Requirement.parse(req)
  File "/galaxy/.venv/local/lib/python2.7/site-packages/pip/_vendor/pkg_resources.py", line 2606, in parse
    reqs = list(parse_requirements(s))
  File "/galaxy/.venv/local/lib/python2.7/site-packages/pip/_vendor/pkg_resources.py", line 2544, in parse_requirements
    line, p, specs = scan_list(VERSION,LINE_END,line,p,(1,2),"version spec")
  File "/galaxy/.venv/local/lib/python2.7/site-packages/pip/_vendor/pkg_resources.py", line 2522, in scan_list
    "Expected ',' or end-of-list in",line,"at",line[p:]
ValueError: ("Expected ',' or end-of-list in", 'pip==8.0.2+gx1', 'at', '+gx1')
```

nsoranzo commented:
I agree with @martenson, when starting from an old pip (e.g. 1.5.4 shipped by Ubuntu 14.04.3 LTS) I can replicate this, the plus sign in the version is not supported.

Possible workaround which seems to work for me:

```
pip install $pre --upgrade https://wheels.galaxyproject.org/packages/pip-8.0.2+gx1-py2.py3-none-any.whl
```

natefoo commented:
@nsoranzo Unfortunately this will reinstall pip every time instead of skipping since it knows it is already installed:

``` console
(test)nate@weyerbacher% PYTHONPATH= pip install --pre --upgrade https://wheels.galaxyproject.org/packages/pip-8.0.2+gx1-py2.py3-none-any.whl
Collecting pip==8.0.2+gx1 from https://wheels.galaxyproject.org/packages/pip-8.0.2+gx1-py2.py3-none-any.whl
  Downloading https://wheels.galaxyproject.org/packages/pip-8.0.2+gx1-py2.py3-none-any.whl (1.2MB)
    100% |████████████████████████████████| 1.2MB 26.2MB/s 
Installing collected packages: pip
  Found existing installation: pip 7.1.2
    Uninstalling pip-7.1.2:
      Successfully uninstalled pip-7.1.2
Successfully installed pip-8.0.2+gx1
(test)nate@weyerbacher% PYTHONPATH= pip install --pre --upgrade https://wheels.galaxyproject.org/packages/pip-8.0.2+gx1-py2.py3-none-any.whl
Collecting pip==8.0.2+gx1 from https://wheels.galaxyproject.org/packages/pip-8.0.2+gx1-py2.py3-none-any.whl
  Downloading https://wheels.galaxyproject.org/packages/pip-8.0.2+gx1-py2.py3-none-any.whl (1.2MB)
    100% |████████████████████████████████| 1.2MB 41.9MB/s 
Installing collected packages: pip
  Found existing installation: pip 8.0.2+gx1
    Uninstalling pip-8.0.2+gx1:
      Successfully uninstalled pip-8.0.2+gx1
Successfully installed pip-8.0.2+gx1
(test)nate@weyerbacher% PYTHONPATH= pip install --pre --upgrade pip==8.0.2+gx1
Requirement already up-to-date: pip==8.0.2+gx1 in /home/nate/.virtualenvs-stretch/test/lib/python2.7/site-packages
```

As we do with `--pre` I could figure out what the first version supporting `+` is and use the URL syntax on anything older. Other ideas welcome.

nsoranzo commented:
@natefoo You're obviously right, didn't thought about that! Your proposed solution seems the way to go, and also reuses my idea, so I'm happy :)

nsoranzo commented:
pip 6.0 seems to be the first release supporting `+`.

natefoo committed:
pip < 1.6 cannot parse requirement specifiers to the `install` command
that contain a `+` in the version. For these older versions, directly
install the wheel or install from the sdist. Technically 1.4 and later
can install wheels, but in testing, only 1.5+ was able to install them
properly with their scripts.

The new heredoc execute and redirect syntax is Bourne-compatible.

pip --pre is no longer necessary since the Galaxy pip version is no
longer a development version of pip.

natefoo commented:
Thanks for checking the version for me @nsoranzo. I've updated the PR with something that should work back to at least pip 1.0.

nsoranzo commented on scripts/common_startup.sh:
```diff
@@ -137,8 +139,28 @@ fi
+    method=`python - << EOF
+from pkg_resources import parse_version
+from sys import stdout
+if parse_version('$pip_version') >= parse_version('1.6'):
```
Should this be 6.0 ? There is no 1.6 release. Or have you tested some 1.6 pre-release?

natefoo replied:
Should indeed be - it all worked as it was since there is no version >= 1.6 and < 6.0

nsoranzo commented on scripts/common_startup.sh:
```diff
@@ -137,8 +139,28 @@ fi
+EOF`
+    case $method in
+        req)
+            pip install --no-index --find-links ${GALAXY_WHEELS_INDEX_URL}/pip --upgrade "pip==${GXPIP_VERSION}"
```
Is `--pre` not needed anymore?

natefoo replied:
It is not, since we are not installing a prerelease version of pip.

natefoo committed:
pip 1.6 -> 6.0, thanks @nsoranzo

nsoranzo commented:
:+1:
